### PR TITLE
[WIP] Updating CMakeLists.txt to use CPM for Package Management and / take advantage of cmake 3.18 support.

### PR DESCRIPTION
**Pull request by millerhooks**
In an attempt to wrangle `RAPIDS` `CMakeLists.txt` I've adopted the use of `CPM` as a package manager and audited some other issues. This PR is not meant to dictate where to go with `cmake` but just an attempt to solve a few problems and get some discussion going about solving problems going forward. Just to try to streamline that discussion I've outlined what I'm doing and why below as well as some excerpts from the `CPM` documentation. 

My tear down and rebuild process for `cudf` was all in the context of `rapids-compose` so I could easily destroy `ccache` and `conda` to verify multiple states. I do not think this PR fully reflects the MVP to merge this pattern. 

I also did a treatment using the package manager `hunter` instead of `CPM` (Draft PR #6026) . I found that `hunter` was more ops focused and had a pattern focused on building a repository of built ccache binaries. `CPM` is a much simpler project and seems to fit our goals better.

CMakeTests.txt Outline
==================
I drew this little thing to give an overview of the existing `CMakeLists.txt`, it's functions, and what is modified when upgrading to cmake 3.18 and using CPM. I numbered the sections to make it easier to discuss in this PR.

CUDF-CMAKE-Upgrade-Overview

Problems and Goals
================== 

Problems
--------
* Our `cmake` (in all `RAPIDS` projects) has accumulated some neglected copypasta that bogs down production and steals productivity 
* We cannot upgrade to the newest `cmake` (3.18) which has very useful helpers for `CUDA` support because of issues relating to our complicated environment flow
* Not having a simple, standards driven `cmake` causes repetition of bad but functional tooling
* `cmake` is an afterthought for many developers slowing down production when `cmake` ontology is not clear
* New `cmake 3.18` `CUDA` features are broken when we do our `conda` environment steps in our `CMakeList.txt`
  - Do we still need our `CUDA` architecture detection?
  - In some environments [`rapids-compose`] does our current `conda` environment break the new (`cmake 3.18`) `nvcc` / `cuda` detection?
    * `/conda/cuda_10.2/envs/rapids/compiler_compat/ld: cannot find -lcudf`

Goals
-----
* Make `RAPIDS` `cmake` compatible with `v3.18` to utilize new native `CUDA` features
* Do not build a new `cmake` package management paradigm using `CPM`. 
* Decrease `RAPIDS` install conflicts related to existing system installed packages
* Pair down redundant and complicated environment vars, includes, and linked directories.
  - Standardize target linking to enable using `PUBLIC` and `PRIVATE` instead of passing around package name variables.
* Speed up build times
* Increase resilience and predictability to dependency management
* A single source of truth for future configurations in `RAPIDS` projects
* Easy incremental integration without disrupting development
* Easy CI/CD integration
* Easy improvements, enhancements, and ability to refactor for `cmake` configuration for all rapids projects
* Reduce malignant duplication between `RAPIDS` projects `cmake` configurations

Overview of Changes
===================
* Adding `CPM` as the `cmake` package manager.
  - https://github.com/TheLartians/CPM.cmake
* Simplifying `zlib`, `boost-filesystem`, `gtest`, and `benchmarks` management via `CPM`
* Adding `gtest` and `benchmarks` to the singluar `target_link_libraries`
  - I am unsure about the universality of `${<package>_LIBRARIES}` usage but hopeful that it can work as a model for single point collection
* [WIP] Using new `FindCUDAToolkit` features of `cmake 3.18` to simplify GPU architecture detection and loading of `cudart` packages.
  - Currently the detection of the package is failing, though it seems to be able to identify the installed `CUDA` version.
* [TODO] Build `arrow` with `CPM` falling back to system if correct options are flagged, if not build from source.

I have not directly addressed but am aware of @jdye64 's outline for changes to our cmake configuration. Some of his issues are partially addressed and I believe we can encapsulate the issues he's brought up in this push to adopt 3.18 and CPM.

Ask @jdye64 for a link if you're curious about the document. Issues #5940 and #5569 are two thirds of it.

Testing Procedure
=================
To properly test this you should use Paul Taylor's [@trxcllnt] `rapids-compose` project [github.com/trxcllnt/rapids-compose] and not your system directly. This will make it easier to modify environment to understand all the variable involved and also not waste too much time and possibly upset your normal development environment.

About CPM
==========

Comparison to pure FetchContent / ExternalProject
--------------------------------------------------------------------
https://github.com/TheLartians/CPM.cmake#comparison-to-pure-fetchcontent--externalproject

CPM.cmake is a wrapper for CMake's FetchContent module and adds a number of features that turn it into a useful dependency manager. The most notable features are:

* A simpler to use API
* Version checking: CPM.cmake will check the version number of any added dependency and omit a warning if another dependency requires a more recent version.
* Options: any Options passed to a dependency are stored and compared on later use, so if another dependency tries to add an existing dependency with incompatible options a warning will be emitted to the user.
* Offline builds: CPM.cmake will override CMake's download and update commands, which allows new builds to be configured while offline if all dependencies are available locally.
* Automatic shallow clone: if a version tag (e.g. v2.2.0) is provided and CPM_SOURCE_CACHE is used, CPM.cmake will perform a shallow clone of the dependency, which should be significantly faster while using less storage than a full clone.
* Overridable: all CPMAddPackage can be configured to use find_package by setting a CMake flag, making it easy to integrate into projects that may require local versioning through the system's package manager.
* Package lock files for easier transitive dependency management.
* Dependencies can be overridden per-build using CMake CLI parameters.
* ExternalProject works similarly as FetchContent, however waits with adding dependencies until build time. This has a quite a few disadvantages, especially as it makes using custom toolchains / cross-compiling very difficult and can lead to problems with nested dependencies.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by kkraus14** [comment]
Writeup looks great. Haven't had a chance to dig into the changes yet, but one comment is that even for header only libraries like Thrust / CUB / DLPack / etc. We can still use `CPM` to only download them if desired, but in the case of these three libraries they expose targets even though they're header only, and we can / should use those targets to automatically handle include directories.

Similarly, we should be able to use CPM to find or build RMM similarly to the other projects as well. We just need to pay special attention to allowing a user to "install" by just setting `LD_LIBRARY_PATH` to the folder with the `librmm.so` (as long as RMM isn't header only, and once we similarly enable exporting the cmake info from an RMM build then we can use targets to handle RMM as well.

**Comment by jdye64** [comment]
Looks good @millerhooks Does `CPM` offer a way to "export" the targets it generates? For example if a third-party application wanted to use `libcudf` would they need to include all the dependencies again or would they already be exported in the libcudf target?

File: cpp/cmake/CPM.cmake [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+elseif(DEFINED ENV{CPM_SOURCE_CACHE})
+  set(CPM_DOWNLOAD_LOCATION "$ENV{CPM_SOURCE_CACHE}/cpm/CPM_${CPM_DOWNLOAD_VERSION}.cmake")
+else()
+  set(CPM_DOWNLOAD_LOCATION "${CMAKE_BINARY_DIR}/cmake/CPM_${CPM_DOWNLOAD_VERSION}.cmake")
+endif()
+
+if(NOT (EXISTS ${CPM_DOWNLOAD_LOCATION}))
+  message(STATUS "Downloading CPM.cmake to ${CPM_DOWNLOAD_LOCATION}")
+  file(DOWNLOAD
+       https://github.com/TheLartians/CPM.cmake/releases/download/v${CPM_DOWNLOAD_VERSION}/CPM.cmake
```
[the remark this answers is not in the document]
Comment (dantegd) [reply cpp/cmake/CPM.cmake:14]: please lets vendor this indeed, this is a component that could cause bad headaches if we don't!

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -159,6 +160,20 @@ option(BUILD_SHARED_LIBS "Build shared libraries" ON)
...
 
+###################################################################################################
+# - CPM options --------------------------------------------------------------------------------
+
+# Setting cache directory for `compose`'s `.ccache` directory. The default location (/etc/.cache) has 
+# some issues with permissions. This is not sustainable and should not be merged.
+# It's here to get this running so we can talk about it. I need some guidance on the prefered acceptable
+# location for a cache. Should it be `thirdparty`?
+# https://github.com/TheLartians/CPM.cmake
+set(CPM_SOURCE_CACHE "$ENV{HOME}/compose/etc/.ccache/CPM")
```
[the remark this answers is not in the document]
Comment (millerhooks) [reply cpp/CMakeLists.txt:171]: Yes. I agree and why I put the verbose comment in. `.rapids` seems like a good choice, I wasn't quite sure where it should end up. Thanks!

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -256,8 +277,18 @@ file(INSTALL ${LIBCXX_INCLUDE_DIR}     DESTINATION ${CMAKE_BINARY_DIR}/include/l
...
+find_package(ZLIB REQUIRED QUIET)
+
+if (NOT ZLIB_FOUND)
+  CPMFindPackage(
+    NAME ZLIB
+    GITHUB_REPOSITORY madler/zlib
+    GIT_TAG 1.2.11
+    VERSION 1.2.11
+  )
+endif()
```
[the remark this answers is not in the document]
Comment (millerhooks) [reply cpp/CMakeLists.txt:289]: Paraphrasing my comments from L389

```
  # This method detects if it's local, but if it is not available uses CPM to install the package.
  # By default CPM does not add the standard variables that come with find package such as `GTEST_LIBRARY_DIR`.
  # So we detect then use `CPMFind/AddPackage` and set the library dir ourselves. 
  # Possibly this is worth patching to support but I don't fully grok the design decision yet.
```
We should not need to, but at the phase I'm in I've been wading through import and linking strategies. This is here so I can test out this pattern with a know good library that is importing and building properly and I can verify the linking. But yes, you are correct. I'm 90% of the way to adding a wrapper for CPM to set standard vars for imported kit.

While typing this I realize I didn't link to the docs on that and possibly the variables are set with Find but not Add. Will check and solve. 

In summary; yes, you are correct this is not what should be done.

File: cpp/CMakeLists.txt [hunk 4]
```diff
@@ -276,8 +307,20 @@ set(Boost_NO_BOOST_CMAKE ON)
...
     COMPONENTS filesystem
+    QUIET
 )
 
+if (NOT Boost_FOUND)
+  CPMFindPackage(
+    NAME Boost
+    GITHUB_REPOSITORY Orphis/boost-cmake
+    GIT_TAG 1.70.0
+    VERSION 1.70.0
```
[the remark this answers is not in the document]
Comment (millerhooks) [reply cpp/CMakeLists.txt:318]: Ok. I'll dig and figure out what we require.

[the remark this answers is not in the document]
Comment (kkraus14) [reply cpp/CMakeLists.txt:318]: I don't think we have a strong version dependency, so we typically follow the conda forge pinning which is currently 1.72.0.

File: cpp/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkraus14) [reply cpp/CMakeLists.txt:289]: Got it. Yea this is because doing a `find_package(ZLIB)` hits a CMake specific `FindZLib` module: https://cmake.org/cmake/help/v3.18/module/FindZLIB.html

File: cpp/CMakeLists.txt [hunk 5]
```diff
@@ -312,7 +356,7 @@ endif(RMM_INCLUDE AND RMM_LIBRARY)
 
 find_path(
     DLPACK_INCLUDE "dlpack"
-    HINTS "$ENV{DLPACK_ROOT}/include"
+    HINTS "$ENV{DLPACK_ROOT}/inclZLIB_LIBRARIESude"
```
Comment (germasch) [inline cpp/CMakeLists.txt:359]: This definitely looks like a typo ;)

**Comment by germasch** [comment]
I'm doing somewhat related work in for rmm, and obviously it makes sense to keep things consistent throughout rapids.

My main goal is somewhat contrary to the primary goal here, I guess, in that I'd like to make rmm usable from other cmake projects via the "usual" mechanism, ie., by `find_package(rmm)` which relies on a cmake config installed together with rmm.

Where it overlaps is that installing rmm the usual way isn't easily possible, because rmm relies on other vendored libraries (e.g., spdlog), which don't get installed together with rmm -- and they shouldn't really. So what I need is a way to build rmm against pre-installed libraries, rather than relying on all the vendored versions that are pulled in via `FetchContent`, etc.

I wasn't familiar with CPM, but it looks like it's a solution for both use cases -- it'll make it easier to build rapids libraries from scratch where all dependencies are automatically downloaded and installed, but it also supports the use case of using pre-existing libraries, which is essentially what's needed when trying to package rmm using an external package manager like conda or spack (the latter being my target, but it doesn't really matter).

So I think this looks like a great way to support both vendored and pre-installed dependencies, and I'll give that a shot for rmm (I'm sure, as usual, the devil's in the details...)

**Comment by kkraus14** [comment]
Thanks @germasch! We decided to give this a shot with cuDF first as it's much more complicated / involved, so we could evaluate CPM as a solution. One of the goals with this work is to also allow adding RMM itself via CPM to either find an existing install or add it if it's not found.

File: cpp/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (germasch) [reply cpp/CMakeLists.txt:289]: Not quite related, but this code snippet won't work as intended because of the `REQUIRED` in the `find_package()`: If zlib isn't found, the config will error out right there.

**Comment by harrism** [comment]
Definitely pushing to 0.17.

**Comment by harrism** [comment]
@kkraus14 should we kick this can down to 0.18?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by kkraus14** [comment]
@harrism I'm going to close this as Miller is no longer working on RAPIDS and this would still need a significant amount of work before it's ready to be merged. I'll likely open new PR(s) in 0.18 tackling this.
